deps/oauth2_client/src/oauth2_client.erl

```diff
@@ -0,0 +1,476 @@
+                  {ok, OAuthProvider2};
+                _ = Attrs->
+                  {error, {missing_oauth_provider_attributes, Attrs}}
+              end;
+          {error, _} = Error3 -> Error3
+        end
+  end.
+
+
+-spec get_oauth_provider(oauth_provider_id(), list()) -> {ok, oauth_provider()} | {error, any()}.
+get_oauth_provider(OAuth2ProviderId, ListOfRequiredAttributes) when is_list(OAuth2ProviderId) ->
+  get_oauth_provider(list_to_binary(OAuth2ProviderId), ListOfRequiredAttributes);
+
+get_oauth_provider(OAuth2ProviderId, ListOfRequiredAttributes) when is_binary(OAuth2ProviderId) ->
+  rabbit_log:debug("get_oauth_provider ~p with at least these attributes: ~p", [OAuth2ProviderId, ListOfRequiredAttributes]),
+  case lookup_oauth_provider_config(OAuth2ProviderId) of
+    {error, _} = Error0 ->
+    rabbit_log:debug("Failed to find oauth_provider ~p configuration due to ~p", [OAuth2ProviderId, Error0]),
+      Error0;
+    Config ->
+      rabbit_log:debug("Found oauth_provider configuration ~p", [Config]),
+      OAuthProvider = case Config of
+        {error,_} = Error -> Error;
+        _ -> map_to_oauth_provider(Config)
+      end,
+      rabbit_log:debug("Resolved oauth_provider ~p", [OAuthProvider]),
+      case find_missing_attributes(OAuthProvider, ListOfRequiredAttributes) of
+        [] -> {ok, OAuthProvider};
+        _ -> Result2 = case OAuthProvider#oauth_provider.issuer of
+                undefined -> {error, {missing_oauth_provider_attributes, [issuer]}};
+                Issuer ->
+                    rabbit_log:debug("Downloading oauth_provider ~p using issuer ~p", [OAuth2ProviderId, Issuer]),
+                    case get_openid_configuration(Issuer, get_ssl_options_if_any(OAuthProvider)) of
+                      {ok, OauthProvider} -> {ok, update_oauth_provider_endpoints_configuration(OAuth2ProviderId, OauthProvider)};
+                      {error, _} = Error2 -> Error2
+                    end
+            end,
+            case Result2 of
+              {ok, OAuthProvider2} ->
+                  case find_missing_attributes(OAuthProvider2, ListOfRequiredAttributes) of
+                    [] ->
+                      rabbit_log:debug("Resolved oauth_provider ~p", [OAuthProvider]),
+                      {ok, OAuthProvider2};
+                    _ = Attrs->
+                      {error, {missing_oauth_provider_attributes, Attrs}}
+                  end;
+              {error, _} = Error3 -> Error3
+            end
+      end
+  end.
+
+%% HELPER functions
+
+
+oauth_provider_to_proplists(#oauth_provider{} = OAuthProvider) ->
+  lists:zip(record_info(fields, oauth_provider), tl(tuple_to_list(OAuthProvider))).
+filter_undefined_props(PropList) ->
+  lists:foldl(fun(Prop, Acc) ->
+    case Prop of
+      {Name, undefined} -> Acc ++ [Name];
+      _ -> Acc
+    end end, [], PropList).
+
+-spec intersection(list(), list()) -> list().
+intersection(L1, L2) ->
```

Comment:
I can see two intersection functions in the code but with different impl. Worth settling on one?